### PR TITLE
PLANNER-2269: Make CS-B Tuple a plain data carrier

### PR DESCRIPTION
**Pull request by triceo**

### CODE REVIEW

File: optaplanner-core/src/main/java/org/optaplanner/core/impl/score/stream/bavet/BavetConstraintSession.java
Comment (triceo) [inline optaplanner-core/src/main/java/org/optaplanner/core/impl/score/stream/bavet/BavetConstraintSession.java:81]: This is not actually in use yet. The change to completely remove `Node` reference from `Tuple` seemed too invasive for this PR.

File: optaplanner-core/src/main/java/org/optaplanner/core/impl/score/stream/bavet/common/BavetTupleState.java
Comment (triceo) [inline optaplanner-core/src/main/java/org/optaplanner/core/impl/score/stream/bavet/common/BavetTupleState.java:20]: Why is `NEW` not `dirty`?

File: optaplanner-core/src/main/java/org/optaplanner/core/impl/score/stream/bavet/BavetConstraintSession.java
Comment (ge0ffrey) [inline optaplanner-core/src/main/java/org/optaplanner/core/impl/score/stream/bavet/BavetConstraintSession.java:59]: I guess this can be `nodeSize`? (the number of nodes)

File: optaplanner-core/src/main/java/org/optaplanner/core/impl/score/stream/bavet/BavetConstraintSession.java
Comment (ge0ffrey) [reply optaplanner-core/src/main/java/org/optaplanner/core/impl/score/stream/bavet/BavetConstraintSession.java:81]: Yes, thank you

**Comment by triceo** [comment]
Pre-requisite: #1064 
Will cause merge conflicts here, will have to resolve.

**Comment by triceo** [comment]
I have rebased this on top of #1064. When that is merged, I will rebase this again on `master`.

**Comment by triceo** [comment]
Rebased, making this PR much smaller.

**Comment by triceo** [comment]
The performance issue was resolved by reverting some of the changes without which we can live.

File: optaplanner-core/src/main/java/org/optaplanner/core/impl/score/stream/bavet/BavetConstraintSession.java
Comment (ge0ffrey) [inline optaplanner-core/src/main/java/org/optaplanner/core/impl/score/stream/bavet/BavetConstraintSession.java:93]: Why doesn't getNode() return type BavetAbstractNode?

File: optaplanner-core/src/main/java/org/optaplanner/core/impl/score/stream/bavet/BavetConstraintSession.java
Comment (ge0ffrey) [inline optaplanner-core/src/main/java/org/optaplanner/core/impl/score/stream/bavet/BavetConstraintSession.java:105]: Let's a default case that throws an exception?

Comment (ge0ffrey) [reply optaplanner-core/src/main/java/org/optaplanner/core/impl/score/stream/bavet/BavetConstraintSession.java:105]: Also, this exception can start with "Impossible state:"
```suggestion
                        "Impossible state: The tuple (" + tuple + ") is already in the dead state (" + tuple.getState() + ").");
```

Comment (ge0ffrey) [reply optaplanner-core/src/main/java/org/optaplanner/core/impl/score/stream/bavet/BavetConstraintSession.java:105]: Do we also want to print out the node toString in the exception message? Might be useful if someone ever sees it on productions and asks us what could be the cause.

File: optaplanner-core/src/main/java/org/optaplanner/core/impl/score/stream/bavet/bi/BavetFilterBiNode.java
Comment (ge0ffrey) [inline optaplanner-core/src/main/java/org/optaplanner/core/impl/score/stream/bavet/bi/BavetFilterBiNode.java:50]: (important IMO)
The entire Bavet implementation is not public API. We don't need to wrap in `unmodifiableList()` etc, as that will only create perf loss and memory usage increase.

File: optaplanner-core/src/main/java/org/optaplanner/core/impl/score/stream/bavet/bi/BavetGroupBiNode.java
Comment (ge0ffrey) [inline optaplanner-core/src/main/java/org/optaplanner/core/impl/score/stream/bavet/bi/BavetGroupBiNode.java:46]: For consistency sake, please call this method `getChildNodeList()`.

The color of the curtains is pink, I know you hate it, but let's not mix colors of curtains - it's going to look ugly.

File: optaplanner-core/src/main/java/org/optaplanner/core/impl/score/stream/bavet/bi/BavetGroupBiNode.java
Comment (ge0ffrey) [inline optaplanner-core/src/main/java/org/optaplanner/core/impl/score/stream/bavet/bi/BavetGroupBiNode.java:47]: see above, no need for unmodifiableList

File: optaplanner-core/src/main/java/org/optaplanner/core/impl/score/stream/bavet/bi/BavetJoinBridgeBiTuple.java
Comment (ge0ffrey) [inline optaplanner-core/src/main/java/org/optaplanner/core/impl/score/stream/bavet/bi/BavetJoinBridgeBiTuple.java:31]: This is intresting. The Set becomes a List (which is a good thing I think.)

Did this bring a performance gain? If not, is it possible it was offset by the unmodifiable wrapping?

File: optaplanner-core/src/main/java/org/optaplanner/core/impl/score/stream/bavet/tri/BavetJoinTriNode.java
Comment (ge0ffrey) [inline optaplanner-core/src/main/java/org/optaplanner/core/impl/score/stream/bavet/tri/BavetJoinTriNode.java:55]: see above: method name + unmodifiableList

**Review by ge0ffrey — APPROVED** [review]
LGTM if unmodifiableList are gone

File: optaplanner-core/src/main/java/org/optaplanner/core/impl/score/stream/bavet/bi/BavetJoinBridgeBiTuple.java
Comment (triceo) [reply optaplanner-core/src/main/java/org/optaplanner/core/impl/score/stream/bavet/bi/BavetJoinBridgeBiTuple.java:31]: I'll re-benchmark once this is merged. But I don't really think this should result in such a big gain - it's just inside one particular tuple, not inside all of them.

**Comment by triceo** [comment]
Jenkins retest this.

**Pull request merged**
